src/core/mempool.rs

```diff
@@ -504,11 +538,72 @@ impl MemPoolDB {
         ))
     }
 
+    fn get_next_timestamp_at_next_block<E>(
+        &self,
+        chainstate: &mut StacksChainState,
+        mut curr_tip_consensus_hash: ConsensusHash,
+        mut curr_tip_block_hash: BlockHeaderHash,
+        mut curr_tip_height: Option<u64>,
+        max_tip_consensus_hash: &ConsensusHash,
+        max_tip_block_hash: &BlockHeaderHash,
+        max_tip_height: u64,
+    ) -> Result<Option<MemPoolWalkResult>, E>
+    where
+        E: From<db_error> + From<ChainstateError>,
+    {
+        loop {
+            // walk to where the first transaction we can mine can be found
+            match self.walk_forwards(
+                chainstate,
+                &curr_tip_consensus_hash,
+                &curr_tip_block_hash,
+                curr_tip_height,
+                &max_tip_consensus_hash,
+                &max_tip_block_hash,
+                max_tip_height,
+            )? {
+                MemPoolWalkResult::Chainstate(
+                    next_consensus_hash,
+                    next_block_bhh,
+                    next_height,
+                    next_timestamp,
+                ) => {
+                    return Ok(Some(MemPoolWalkResult::Chainstate(
+                        next_consensus_hash,
+                        next_block_bhh,
+                        next_height,
+                        next_timestamp,
+                    )));
+                }
+                MemPoolWalkResult::NoneAtHeight(
+                    next_consensus_hash,
+                    next_block_hash,
+                    next_height,
+                ) => {
+                    if std::env::var("STACKS_MEMPOOL_BAD_BEHAVIOR") == Ok("1".into()) {
```

Comment:
Should this var being read once, before the beginning of the loop?

Reply:
We only want to check for this when we are in the `NoneAtHeight` case, since this captures the behavior of the old walk algorithm. More details here (https://github.com/blockstack/stacks-blockchain/pull/2392)